llms/mlx_lm/models/mixtral.py

```diff
@@ -174,14 +175,60 @@ def __call__(self, x: mx.array) -> mx.array:
         return y.reshape(orig_shape)
 
 
+class MixtralSparseMoeTrainingBlock(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.hidden_dim = args.hidden_size
+        self.ffn_dim = args.intermediate_size
+        self.num_experts = args.num_local_experts
+        self.num_experts_per_tok = args.num_experts_per_tok
+
+        # gating
+        self.gate = nn.Linear(self.hidden_dim, self.num_experts, bias=False)
+
+        self.experts = [
+            MixtralBLockSparseTop2MLP(args=args) for _ in range(self.num_experts)
+        ]
+
+    def __call__(self, x: mx.array) -> mx.array:
+        ne = self.num_experts_per_tok
+        orig_shape = x.shape
+        x = x.reshape(-1, x.shape[-1])
+
+        gates = self.gate(x)
+        inds = mx.stop_gradient(mx.argpartition(-gates, kth=ne, axis=-1)[:, :ne])
+
+        scores = mx.softmax(
+            mx.take_along_axis(gates, inds, axis=-1).astype(mx.float32),
+            axis=-1,
+        ).astype(gates.dtype)
+
+        mx.eval(inds)
+        inds = np.array(inds)
+        y = mx.zeros((x.shape[0], ne, x.shape[-1]))
+        for e, expert in enumerate(self.experts):
+            idx1, idx2 = map(mx.array, np.where(inds == e))
+            if idx1.size == 0:
+                continue
+            y[idx1, idx2] = expert(x[idx1])
+
+        y = (y * scores[:, :, None]).sum(axis=1)
+
+        return y.reshape(orig_shape)
+
+
 class MixtralDecoderLayer(nn.Module):
     def __init__(self, args: ModelArgs):
         super().__init__()
         self.hidden_size = args.hidden_size
 
         self.self_attn = MixtralAttention(args)
 
-        self.block_sparse_moe = MixtralSparseMoeBlock(args)
+        self.block_sparse_moe = (
+            MixtralSparseMoeTrainingBlock(args)
```

Comment:
For training/inferencing, we use different models layers to avoid performance degradation.

Reply:
It might be cleaner to to use the same __call__ and check the last dimension of `y` just before the expert loop and branch based on that. That way we could also use the batching version prompts. Wdyt?

Reply:
Indeed that’s exactly why I didn’t do it in the first place for inference only. I think it has to do with memory use being different but I didn’t figure it out yet.

but sounds good let’s keep it separate for now until perf issues are debugged.

Reply:
Another alternative is to use the Module.training flag to branch? I am not sure what the implications are for the rest of the code

Reply:
> Another alternative is to use the Module.training flag to branch? I am not sure what the implications are for the rest of the code

I have considered using `model.train()`, but I need to construct different layers during model initialization. So, `model.train` still act as a runtime behavior change that is equivalent to an if-else statement in the forward pass. Maybe I misunderstood.

Reply:
When you notice the perf degradation that was for inference right? The degradation is almost certainly not due to just checking the condition but due to having a different path for prompt processing and single token generation.

If you use model.training to branch instead of the shape of the input in the `__call__` function you will have the same path for prompt processing and token generation (the loop) and a different branch for training (the batching over experts).

Reply:
> When you notice the perf degradation that was for inference right? The degradation is almost certainly not due to just checking the condition but due to having a different path for prompt processing and single token generation.
> 
> If you use model.training to branch instead of the shape of the input in the `__call__` function you will have the same path for prompt processing and token generation (the loop) and a different branch for training (the batching over experts).

Sounds good, I will give it a try. If that works, it would be great. I don't really like the current implementation and adding an additional train flag in model arguments.:)

Reply:
module.training works well. I didn't find any performance impact from my local test. Thank you for pointing out the better solution 🚀